### DISCUSSION TITLE
Different symbols in static and shared library

### DISCUSSION THREAD

**Discussion by GS-JVarju** [openssl/openssl#25950] [Q&A] [answered]
I configured and built openssl 3.0.15 on Windows, and it works fine. I have the static libraries and the shared too. When I tried to link my application to one of these I noticed that it works with the static lib but not with the dll. It seems like the exported symbols are different. Is it intentional? Am I missing something?

I created a dummy application:

```
#include <stdio.h>
#include <crypto/chacha.h>

int main() {
    unsigned char key[32] = {0};
    unsigned char nonce[12] = {0};
    unsigned char in[64] = {0};
    unsigned char out[64] = {0};

    // Initialize the key and nonce with some values
    for (int i = 0; i < 32; i++) key[i] = i;
    for (int i = 0; i < 12; i++) nonce[i] = i;

    // Encrypt the input data
    ChaCha20_ctr32(out, in, sizeof(in), (const unsigned int *)key, (const unsigned int *)nonce);

    // Print the encrypted output
    printf("Encrypted output:\n");
    for (int i = 0; i < sizeof(out); i++) {
        printf("%02x", out[i]);
        if ((i + 1) % 16 == 0) printf("\n");
    }

    return 0;
}
```

And I tried to build it like this:
`
cl /I <path-to-openssl>\openssl-openssl-3.0.15\include linktest.c /link /LIBPATH:<path-to-openssl>\openssl-openssl-3.0.15 libcrypto_static.lib`

This works.

This does not:

`cl /I <path-to-openssl>\openssl-openssl-3.0.15\include linktest.c /link /LIBPATH:<path-to-openssl>\openssl-openssl-3.0.15 libcrypto.lib`

Compiler says this:
```
linktest.obj : error LNK2019: unresolved external symbol ChaCha20_ctr32 referenced in function main
linktest.exe : fatal error LNK1120: 1 unresolved externals
```

**Comment by levitte** [comment, marked as answer]
As far as I know, non-static symbols cannot be hidden in static libraries, and OpenSSL has a number of internal symbols that aren't meant for public consumption, and are therefore not exported in the DLL, where it's possible to have that sort of control.

Notably, the headers in include/openssl are public headers (`nmake install` will copy them to the installation directory), while those in include/crypto and include/internal are internal, i.e. not meant for public consumption.

Strictly speaking, since you are using an internal function, you're pretty much on your own.  ChaCha20_ctr32 may be renamed or removed some time in the future, and as you've observed, it's not accessible with the DLL.  You're better off using the EVP_Cipher interface, which is mainly documented here: https://docs.openssl.org/master/man3/EVP_EncryptInit/

**Comment by mattcaswell** [comment]
I saw your stackoverflow question before I saw your question here. See my answer there:

https://stackoverflow.com/questions/79186122/different-symbols-in-openssl-static-and-shared-library/79188919#79188919

Basically: don't include crypto/chacha.h in your application as @levitte says above.
